### ISSUE TITLE
TF consumes all available RAM with a particular combination of conv2d, batch_norm and LSTM

### ISSUE THREAD

**Issue by noammor** [tensorflow/tensorflow#16627]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: v1.5.0-0-g37aa430d84 1.5.0
- **Python version**: 3.6.4
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**: 9.0.176 / 7.0.5
- **GPU model and memory**: Tesla V100 (AWS P3), 16GB
- **Exact command to reproduce**: python debug_tf.py

### Describe the problem
When the following code is run on a p3.2xlarge, the python process starts consuming RAM indefinitely, until the entire server RAM is used and the server dies. That's system RAM, not GPU memory. 

It doesn't look it, but the code below is the smallest I could find that produces the bad behavior.

1. Replacing residual_conv with a simple convolution makes the code work (i.e not hang).
2. Reducing the repetition number in `layers.repeat` (to e.g 2) makes the code work.
3. Removing the batch normalization from residual_conv makes the code work.
4. Removing the LSTM makes the code work.
5. The weirdest of all, if I set `is_training=True` in batch_norm instead of `is_training=is_training_var` whose value is set to `True` in the feed_dict, then the code works.

The same code runs successfully on a AWS P2 server with TensorFlow 1.3.

### Source code / logs

```python
import numpy as np
import tensorflow as tf
import tensorflow.contrib.layers as layers
from tensorflow.contrib.framework import arg_scope


def residual_conv(incoming, num_filters, scope, bn=True):
	with tf.variable_scope(scope):
	    input_filters = incoming.get_shape().as_list()[-1]
	    if input_filters != num_filters:
	        incoming = layers.conv2d(incoming, num_filters, scope='adjust_conv')

	    after_conv1 = layers.conv2d(incoming, num_filters)
	    after_conv2 = layers.conv2d(after_conv1, num_filters, normalizer_fn=None, activation_fn=None)

	    net = incoming + after_conv2

	    if bn:
	        net = layers.batch_norm(net)

	    return net

def tf_bilstm(incoming, n_units, name):
	net = incoming

	lstm_f = tf.contrib.rnn.LSTMCell(n_units)
	lstm_b = tf.contrib.rnn.LSTMCell(n_units)

	with tf.variable_scope(name):
	    results, _ = tf.nn.bidirectional_dynamic_rnn(lstm_f, lstm_b, net, 
	                                                 dtype=tf.float32, time_major=True)
	return tf.concat(results, axis=2)

def main():
	x_var = tf.placeholder(dtype=tf.float32, shape=(None, None, None, 512))
	is_training_var = tf.placeholder(dtype=bool)

	net = x_var

	with arg_scope([layers.batch_norm], is_training=is_training_var, decay=0.99, scale=True
	               ), \
	     arg_scope([layers.conv2d], padding='SAME', kernel_size=(3, 3)), \
	     arg_scope([layers.max_pool2d], padding='SAME', kernel_size=(2, 2), stride=(2, 2)):

	    net = layers.repeat(net, 20, residual_conv, 64, scope='block1')
	    net = layers.max_pool2d(net)

	    net = tf.squeeze(net, 2)
	    net = tf.transpose(net, [1, 0, 2])
	    net = tf_bilstm(net, 512, 'lstm1')

	sess = tf.Session()
	sess.run(tf.global_variables_initializer())
	
	X = np.zeros([10, 100, 2, 512])

	print('Right before TF call!')

	a = sess.run(net, {x_var: X, is_training_var: True})

	print(a)


if __name__ == '__main__':
	main()
```

Edited: Simplified the code.

**Comment by nxphi47** [comment]
Can you give the logs of the errors?

I believe layers.batch_norm only accept is_training of type python bool, it does not accept tf.placeholder. So if you really want to do `is_training_var`, you may do so,
`
train_bn = layers.batch_norm(is_training=True) # put them same name, same scope
inference_bn = layers.batch_norm(is_training=False, reuse=True)
bn = tf.cond(is_training_var, lambda: train_bn, lambda: inference_bn)
`

**Comment by noammor** [comment]
There is no output, it just hangs and starts eating up memory.

I just tried replacing `batch_norm` with `layer_norm`, which doesn't have an `is_training` parameter at all - no luck, still eats up all the memory.

It's also worth mentioning that the exact same code runs successfully on an AWS P2 server which has tensorflow 1.3 installed.

**Comment by reedwm** [comment]
I can reproduce on 1.5, but not on the master branch. So, this issue has been fixed, but we should probably identity the commit that fixed it and put it in 1.5.

/CC @ebrevdo, any ideas what commit fixed this?

**Comment by herrmann** [comment]
I've had the same problem with a VQA model (CNN for image inputs, LSTM for text inputs, and merging both, then classifying) and pinned the error down to the merging part (tried using concatenate and multiply). RAM goes up -> swap goes up -> machine hangs. Trying to run the sub-models before merging works fine. Works fine on 1.4.1, but I haven't tried using master yet.

**Comment by jingraham** [comment]
I just encountered a similar bug and a weird workaround when upgrading to 1.5 on a different codebase. While trying to diagnose that problem by injecting various Print ops around the graph I found a bizarre fix: if you wrap the placeholder with a tf.identity() (or Print()) and then feed to that tensor rather than the placeholder, it stops hanging.

I just tried it and, sure enough, replacing the line in the above code

`x_var = tf.placeholder(dtype=tf.float32, shape=(None, None, None, 512))`

with

`x_var = tf.identity(tf.placeholder(dtype=tf.float32, shape=(None, None, None, 512)))`

fixes the memory hang problem for me on a K80 instance with Tensorflow 1.5. And, since that tensor still gets fed, it does not seem to affect downstream computation.

Not sure what is common between the codebases, but for what it's worth the one I am working on also involves a combination of conv layers, a tf.while_loop, and placeholders with multiple unknown dimensions.

**Comment by ebrevdo** [comment]
Could it be a grappler issue?

**Comment by reedwm** [comment]
> Could it be a grappler issue?

/CC @zhangyaobit

**Comment by rmlarsen** [comment]
@ebrevdo nothing obvious comes to mind, but it is possible. Can you try to bisect it internally?

**Comment by noammor** [comment]
I can confirm that @jingraham 's workaround fixes the issue. I ended up using this function, since I need to be able to load the placeholders by name:

```python
def placeholder(**kwargs):
    name = kwargs['name']
    del kwargs['name']

    tf_placeholder = tf.placeholder(**kwargs)
    return tf.identity(tf_placeholder, name=name)
```

**Comment by skye** [comment]
I'm gonna close this as it's fixed in master (and presumably 1.6?) and there's a workaround.
